src/descriptor/create_descriptor.rs

```diff
@@ -0,0 +1,446 @@
+//! Tools for creating Descriptor and witness stack from given scriptpubkey and corresponding
+//! scriptsig and witness.
+//!
+//! Figures out the the type of descriptor based on scriptpubkey, witness and scriptsig.
+//! Outputs a descriptor and witness_stack for all descriptors which can be directly fed into the
+//! interpreter. All script_sig and witness are translated into a single witness stack
+//! Vec<StackElement>
+//! 1) Pkh: translates scriptsig to witness_stack<StackElement>, validates sig and pubkey hash
+//! and outputs Pkh descriptor
+//! 2) Wphk: translates witness to witness_stack<StackElement>, validates sig and pubkey hash
+//! and outputs a Wpkh descriptor
+//! 3) Wsh: pops witness script and checks wsh output hash, translates remaining witness elements
+//! to Vec<StackElement> and outputs a Wsh descriptor. Does not check miniscript inside the descriptor
+//! 4) Bare: translates script_sig to witness_stack<StackElement> and script_pubkey to miniscript
+//! 5) Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+//! witness_stack<StackElement> and redeem script to miniscript. Does not check the miniscript
+//! 6) ShWpkh: Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+//! witness_stack<StackElement> and validates Wpkh sig, pubkey.
+//! 7) ShWsh: Checks witness script hash, pops witness script and converts it to miniscript.
+//! translates the remaining witness to witness_stack<StackElement>
+
+use bitcoin::{self, Script, PublicKey};
+use secp256k1;
+
+use descriptor::Descriptor;
+use miniscript::Miniscript;
+use miniscript::evaluate::{verify_sersig, StackElement};
+use Error;
+use ToPublicKey;
+use secp256k1::{VerifyOnly};
+use bitcoin::blockdata::script::Instruction;
+use bitcoin::blockdata::opcodes;
+use ::Error::BadDescriptor;
+
+/// Helper function for creating StackElement from Push instructions. Special case required for
+/// handling OP_PUSHNUM_1.
+/// Dissatisfied and witness are mapped to Pushbytes instruction in scriptsig and satisfied is mapped
+/// as OP_PUSHNUM_1. Other opcodes are considered Non-standard in bitcoin core.
+/// Miniscript should not use other pushes apart from PUSHNUM_1. This will panic on reciving anything
+/// which is not PUSHBYTES, OR PUSHNUM_1 as other things are expected to happen in Miniscript
+fn instr_to_stackelem(ins: Instruction) -> StackElement {
+    match ins {
+        Instruction::PushBytes(v) => StackElement::from(v.to_vec()),
+        Instruction::Op(opcodes::all::OP_PUSHNUM_1) => StackElement::Satisfied,
+        _ => unreachable!(),
```

Comment:
Why don't we translate empty pushes to `Dissatisfied`?

Also we can't panic on non-push data because we're starting from a scriptsig/witness stack which is user-controlled data. We should return an error instead.

Reply:
`Dissatisfied` is actually translated as PUSHBYTES_0 which converted from `StackElement::from(v.to_vec()` . Will change the unreachable to error.

Reply:
Ah, nice. Can you add a comment above `PushBytes(v)` line?